File: tests/unit/pipeline.test.ts
import { afterEach, describe, expect, it } from 'vitest'
import { compress, decompress } from '../../src/codec/compression/deflate'
import { createRaptorqFountain } from '../../src/codec/fountain/raptorq'
import {
  BYTES_PER_TILE,
  FLAG_COMPRESSED,
  META_MAGIC,
  PROTO_VERSION,
  TYPE_DATA,
  type TransferSettings,
} from '../../src/protocol/constants'
import { parseMetadataFrame } from '../../src/protocol/metadata'
import { sha256Hex } from '../../src/protocol/sha256'
import { decodeFrame } from '../../src/protocol/wire'
import {
  PipelineError,
  prepareTransfer,
  repairFrames,
  transferBytes,
  type PreparedTransfer,
} from '../../src/sender/pipeline'

// Deterministic PRNG (mulberry32) so fixtures and drop sets are reproducible.
function mulberry32(seed: number): () => number {
  let a = seed >>> 0
  return () => {
    a += 0x6d2b79f5
    let t = a
    t = Math.imul(t ^ (t >>> 15), t | 1)
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61)
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296
  }
}

function randomBytes(length: number, seed: number): Uint8Array<ArrayBuffer> {
  const rand = mulberry32(seed)
  const bytes = new Uint8Array(length)
  for (let i = 0; i < length; i++) bytes[i] = Math.floor(rand() * 256)
  return bytes
}

function repeatedByte(byte: number, length: number): Uint8Array<ArrayBuffer> {
  return new Uint8Array(length).fill(byte)
}

function first<T>(arr: readonly T[]): T {
  const v = arr[0]
  if (v === undefined) throw new Error('expected a non-empty array')
  return v
}

const INPUT = { filename: 'fixture.bin', mime: 'application/octet-stream' } as const
type TransferInput = Parameters<typeof prepareTransfer>[0]

// The pipeline keeps its encoder alive for repair generation; dispose after
// each test so wasm memory does not leak across the suite.
const liveEncoders: Array<{ dispose(): void }> = []
afterEach(() => {
  for (const enc of liveEncoders) enc.dispose()
  liveEncoders.length = 0
})

async function makeTransfer(input: TransferInput): Promise<PreparedTransfer> {
  const transfer = await prepareTransfer(input)
  liveEncoders.push(transfer.encoder)
  return transfer
}

async function decodePayload(transfer: PreparedTransfer): Promise<Uint8Array> {
  const decoder = await createRaptorqFountain().createDecoder(
    transfer.info.compressedSize,
    transfer.info.mtu,
  )
  let out: Uint8Array | undefined
  for (const frameBytes of transfer.dataFrames) {
    out = decoder.decode(decodeFrame(frameBytes).payload)
    if (out !== undefined) break
  }
  if (out === undefined) throw new Error('decoder did not complete from all source frames')
  return out
}

function restoreOriginal(transfer: PreparedTransfer, payload: Uint8Array): Uint8Array {
  return decompress({ data: payload, compressed: transfer.info.compressed })
}

async function rejectionCode(run: () => Promise<unknown>): Promise<string> {
  let thrown: unknown
  try {
    await run()
  } catch (e) {
    thrown = e
  }
  if (!(thrown instanceof PipelineError)) {
    throw new Error(`expected PipelineError, got ${String(thrown)}`)
  }
  return thrown.code
}

describe('prepareTransfer', () => {
  it('round-trips a 64 KiB random binary via default settings (1k)', async () => {
    const input = randomBytes(64 * 1024, 0x51a1e)
    const transfer = await makeTransfer({ file: input, ...INPUT })
    const info = transfer.info
    const expected = compress(input)

    expect(info.settings).toEqual({
      bytesPerTile: '1k',
      layout: 'grid4',
      targetFps: 15,
      highRefresh: false,
    })
    expect(info.settings.bytesPerTile).toBe('1k')
    expect(info.totalSize).toBe(64 * 1024)
    expect(info.filename).toBe(INPUT.filename)
    expect(info.mime).toBe(INPUT.mime)
    expect(info.symbolSize).toBe(BYTES_PER_TILE['1k'].mtu)
    expect(info.mtu).toBe(BYTES_PER_TILE['1k'].mtu)
    // Random data is high-entropy: deflate usually declines to shrink it, but
    // we assert against the actual compress() result, never a hardcoded value.
    expect(info.compressed).toBe(expected.compressed)
    expect(info.compressedSize).toBe(expected.data.length)
    expect(info.k).toBeGreaterThan(0)
    expect(info.k * info.symbolSize).toBeGreaterThanOrEqual(info.compressedSize)
    expect(info.dataFrameCount).toBe(info.k)
    expect(info.metaFrameCount).toBe(1)
    expect(info.totalFrames).toBe(info.k + 1)
    expect(transferBytes(info)).toBe(info.compressedSize)

    const payload = await decodePayload(transfer)
    expect(payload.length).toBe(info.compressedSize)
    const restored = restoreOriginal(transfer, payload)
    expect(restored).toEqual(input)
    expect(info.fileSHA256).toMatch(/^[0-9a-f]{64}$/)
    expect(await sha256Hex(input)).toBe(info.fileSHA256)
  })

  it('compresses a 100 KiB repeated-byte file and restores the ORIGINAL bytes', async () => {
    const input = repeatedByte('a'.charCodeAt(0), 100 * 1024)
    const transfer = await makeTransfer({ file: input, ...INPUT })

    expect(transfer.info.compressed).toBe(true)
    expect(transfer.info.compressedSize).toBeLessThan(transfer.info.totalSize)

    const restored = restoreOriginal(transfer, await decodePayload(transfer))
    expect(restored).toEqual(input)
    expect(await sha256Hex(input)).toBe(transfer.info.fileSHA256)
  })

  it('handles a 1-byte file with k >= 1', async () => {
    const input = new Uint8Array([42])
    const transfer = await makeTransfer({ file: input, ...INPUT })

    expect(transfer.info.totalSize).toBe(1)
    expect(transfer.info.k).toBeGreaterThanOrEqual(1)
    expect(restoreOriginal(transfer, await decodePayload(transfer))).toEqual(input)
  })

  it('rejects an empty file with EMPTY_FILE instead of crashing', async () => {
    expect(await rejectionCode(() => makeTransfer({ file: new Uint8Array(0), ...INPUT }))).toBe(
      'EMPTY_FILE',
    )
  })

  it('round-trips a 2 KiB file with 2k (V34) settings', async () => {
    const input = randomBytes(2 * 1024, 77)
    const transfer = await makeTransfer({
      file: input,
      filename: 'v34.bin',
      mime: 'application/octet-stream',
      settings: { bytesPerTile: '2k', layout: 'row3', targetFps: 24, highRefresh: false },
    })

    expect(transfer.info.settings).toEqual({
      bytesPerTile: '2k',
      layout: 'row3',
      targetFps: 24,
      highRefresh: false,
    })
    expect(transfer.info.mtu).toBe(BYTES_PER_TILE['2k'].mtu)
    expect(transfer.info.symbolSize).toBe(transfer.info.mtu)
    expect(transfer.info.k).toBeGreaterThan(0)
    expect(transfer.info.k * transfer.info.mtu).toBeGreaterThanOrEqual(transfer.info.compressedSize)
    expect(restoreOriginal(transfer, await decodePayload(transfer))).toEqual(input)
  })

  it('round-trips a 2 KiB file with 2.5k (V40) settings', async () => {
    const input = randomBytes(2 * 1024, 99)
    const transfer = await makeTransfer({
      file: input,
      filename: 'v40.bin',
      mime: 'application/octet-stream',
      settings: { bytesPerTile: '2.5k', layout: 'single', targetFps: 15, highRefresh: false },
    })

    expect(transfer.info.settings).toEqual({
      bytesPerTile: '2.5k',
      layout: 'single',
      targetFps: 15,
      highRefresh: false,
    })
    expect(transfer.info.mtu).toBe(BYTES_PER_TILE['2.5k'].mtu)
    expect(transfer.info.symbolSize).toBe(transfer.info.mtu)
    expect(transfer.info.k).toBeGreaterThan(0)
    expect(transfer.info.k * transfer.info.mtu).toBeGreaterThanOrEqual(transfer.info.compressedSize)
    expect(restoreOriginal(transfer, await decodePayload(transfer))).toEqual(input)
  })

  it('rejects settings with an unknown bytesPerTile', async () => {
    const settings = {
      bytesPerTile: '9k',
      layout: 'grid4',
      targetFps: 15,
      highRefresh: false,
    } as unknown as TransferSettings
    await expect(makeTransfer({ file: randomBytes(1024, 1), ...INPUT, settings })).rejects.toThrow(
      TypeError,
    )
  })
})

describe('dataFrames', () => {
  it('every DATA frame validates and carries consistent header fields', async () => {
    const transfer = await makeTransfer({ file: randomBytes(64 * 1024, 0x51a1e), ...INPUT })
    const info = transfer.info
    const expectedFlags = info.compressed ? FLAG_COMPRESSED : 0

    for (const frameBytes of transfer.dataFrames) {
      const frame = decodeFrame(frameBytes)
      expect(frame.type).toBe(TYPE_DATA)
      expect(frame.sessionId).toBe(info.sessionId)
      expect(frame.k).toBe(info.k)
      expect(frame.totalLen).toBe(info.compressedSize)
      expect(frame.flags).toBe(expectedFlags)
      expect(frame.payload.length).toBe(info.symbolSize)
    }

    const esis = transfer.dataFrames.map((fb) => decodeFrame(fb).esi)
    expect(new Set(esis).size).toBe(esis.length)
    for (const esi of esis) expect(esi).toBeLessThan(info.k)
    expect(transfer.dataFrames.length).toBe(info.k)
  })
})

describe('metaFrames', () => {
  it('the META frame parses and matches every info field', async () => {
    const transfer = await makeTransfer({ file: randomBytes(16 * 1024, 5), ...INPUT })
    const info = transfer.info

    expect(transfer.metaFrames).toHaveLength(1)
    const meta = parseMetadataFrame(first(transfer.metaFrames))
    expect(meta).toEqual({
      magic: META_MAGIC,
      protoVer: PROTO_VERSION,
      sessionId: info.sessionId,
      filename: info.filename,
      mime: info.mime,
      totalSize: info.totalSize,
      compressedSize: info.compressed ? info.compressedSize : 0,
      compressed: info.compressed,
      k: info.k,
      symbolSize: info.symbolSize,
      mtu: info.mtu,
      fileSHA256: info.fileSHA256,
      flags: info.compressed ? FLAG_COMPRESSED : 0,
    })

    expect(parseMetadataFrame(first(transfer.metaFrames))).toEqual(meta)
    expect(decodeFrame(first(transfer.dataFrames)).sessionId).toBe(meta.sessionId)
  })
})

describe('loss recovery + repair frames', () => {
  it('reassembles the file after losing 20% of source frames', async () => {
    const input = randomBytes(64 * 1024, 0x51a1e)
    const transfer = await makeTransfer({ file: input, ...INPUT })

    // Drop every 5th source frame (~20%), deterministic by construction.
    const survivors = transfer.dataFrames.filter((_, i) => i % 5 !== 0)
    expect(survivors.length).toBeLessThan(transfer.info.k)

    const decoder = await createRaptorqFountain().createDecoder(
      transfer.info.compressedSize,
      transfer.info.mtu,
    )
    let out: Uint8Array | undefined
    for (const frameBytes of survivors) {
      out = decoder.decode(decodeFrame(frameBytes).payload)
      if (out !== undefined) break
    }
    expect(out).toBeUndefined() // survivors alone cannot complete

    for (const frameBytes of repairFrames(transfer, transfer.info.k)) {
      out = decoder.decode(decodeFrame(frameBytes).payload)
      if (out !== undefined) break
    }
    if (out === undefined) throw new Error('decoder did not complete after repair frames')
    expect(restoreOriginal(transfer, out)).toEqual(input)
  })

  it('repairFrames emits valid DATA frames with distinct esi at or beyond k', async () => {
    const transfer = await makeTransfer({ file: randomBytes(16 * 1024, 5), ...INPUT })
    const info = transfer.info
    const expectedFlags = info.compressed ? FLAG_COMPRESSED : 0

    const repairs = repairFrames(transfer, 5)
    expect(repairs).toHaveLength(5)
    const esis = repairs.map((fb) => decodeFrame(fb).esi)
    expect(new Set(esis).size).toBe(5)
    for (const frameBytes of repairs) {
      const frame = decodeFrame(frameBytes)
      expect(frame.type).toBe(TYPE_DATA)
      expect(frame.sessionId).toBe(info.sessionId)
      expect(frame.k).toBe(info.k)
      expect(frame.totalLen).toBe(info.compressedSize)
      expect(frame.flags).toBe(expectedFlags)
      expect(frame.esi).toBeGreaterThanOrEqual(info.k)
    }
  })
})

describe('fileSHA256', () => {
  it('is stable for the same bytes and differs across files', async () => {
    const file = randomBytes(4 * 1024, 7)
    const other = randomBytes(4 * 1024, 8)

    const a = await makeTransfer({ file, ...INPUT })
    const b = await makeTransfer({ file, ...INPUT })
    expect(a.info.fileSHA256).toBe(b.info.fileSHA256)

    const c = await makeTransfer({ file: other, ...INPUT })
    expect(c.info.fileSHA256).not.toBe(a.info.fileSHA256)
  })
})

File: flutter_app/core/test/sender/pipeline_test.dart
// Sender pipeline tests (Wave 3 T3.3) — TDD against a FAKE deterministic
// FountainFactory, so core stays FFI-free.
//
// The fake encoder mirrors the landed fountain interface contract: source
// symbols are deterministic slices of the payload (esi 0..k-1), and
// encodeRepair returns the K source symbols followed by `count` repair
// symbols (esi k..k+count-1). The fake decoder concatenates fed symbols and
// truncates to the transfer's total length, like the real codec.
import 'dart:math';
import 'dart:typed_data';

import 'package:qr_transfer_core/codec/deflate.dart';
import 'package:qr_transfer_core/codec/fountain/interface.dart';
import 'package:qr_transfer_core/protocol/constants.dart';
import 'package:qr_transfer_core/protocol/metadata.dart';
import 'package:qr_transfer_core/protocol/sha256.dart';
import 'package:qr_transfer_core/protocol/wire.dart';
import 'package:qr_transfer_core/sender/pipeline.dart';
import 'package:test/test.dart';

/// Deterministic incompressible bytes (seeded, so tests are repeatable).
Uint8List randomBytes(int length, [int seed = 42]) {
  final rng = Random(seed);
  final bytes = Uint8List(length);
  for (var i = 0; i < length; i++) {
    bytes[i] = rng.nextInt(256);
  }
  return bytes;
}

/// Deterministic fake encoder: symbol i carries payload[i*symbolSize, ...]
/// zero-ish padded to symbolSize; repairs are fresh symbols esi >= k.
class _FakeEncoder implements FountainEncoder {
  _FakeEncoder({
    required this.symbolSize,
    required this.sourceSymbolCount,
    required this.payload,
  });

  @override
  final int symbolSize;

  @override
  final int sourceSymbolCount;

  final Uint8List payload;

  @override
  List<EncodedSymbol> encodeSourceSymbols() => _symbols(0, sourceSymbolCount);

  @override
  List<EncodedSymbol> encodeRepair(int count) => <EncodedSymbol>[
    ..._symbols(0, sourceSymbolCount),
    ..._symbols(sourceSymbolCount, count),
  ];

  List<EncodedSymbol> _symbols(int esiStart, int n) => List.generate(n, (i) {
    final esi = esiStart + i;
    final offset = esi * symbolSize;
    final bytes = Uint8List(symbolSize);
    for (var j = 0; j < symbolSize; j++) {
      final src = offset + j;
      bytes[j] = src < payload.length ? payload[src] : (esi * 31 + j) & 0xff;
    }
    return EncodedSymbol(bytes: bytes, esi: esi);
  });

  @override
  void dispose() {}
}

/// Fake decoder: concatenates fed symbols, returns the truncation to
/// [totalLength] once enough bytes have arrived.
class _FakeDecoder implements FountainDecoder {
  _FakeDecoder(this.totalLength);

  final int totalLength;
  final List<int> _buffer = <int>[];
  bool _complete = false;

  @override
  Uint8List? decode(Uint8List symbolBytes) {
    if (!_complete) {
      _buffer.addAll(symbolBytes);
      _complete = _buffer.length >= totalLength;
    }
    if (!_complete) return null;
    return Uint8List.fromList(_buffer.take(totalLength).toList());
  }

  @override
  bool get isComplete => _complete;

  @override
  void dispose() {}
}

/// Deterministic fake factory with optional knobs to probe the pipeline's
/// error paths: [symbolSizeOverride] (FRAME_TOO_LARGE) and
/// [forceSourceSymbolCount] (K_SANITY_FAILED).
class _FakeFactory implements FountainFactory {
  _FakeFactory({this.symbolSizeOverride, this.forceSourceSymbolCount});

  final int? symbolSizeOverride;
  final int? forceSourceSymbolCount;

  @override
  Future<FountainEncoder> createEncoder(Uint8List data, int mtu) async {
    final symbolSize = symbolSizeOverride ?? mtu - 4;
    final k = forceSourceSymbolCount ?? symbolCountForLength(data.length, mtu);
    return _FakeEncoder(
      symbolSize: symbolSize,
      sourceSymbolCount: k,
      payload: data,
    );
  }

  @override
  Future<FountainDecoder> createDecoder(int totalLength, int mtu) async =>
      _FakeDecoder(totalLength);
}

Matcher pipelineError(PipelineErrorCode code) =>
    isA<PipelineError>().having((e) => e.code, 'code', code);

void main() {
  group('default settings (incompressible input)', () {
    final file = randomBytes(5000);

    test(
      'TransferInfo carries the PWA defaults and derived codec fields',
      () async {
        final prepared = await prepareTransfer(
          file: file,
          filename: 'sample.bin',
          mime: 'application/octet-stream',
          factory: _FakeFactory(),
        );
        final info = prepared.info;

        expect(info.settings.bytesPerTile, BytesPerTileId.oneK);
        expect(info.settings.layout, LayoutId.grid4);
        expect(info.settings.targetFps, 15);
        expect(info.settings.highRefresh, isFalse);
        expect(info.bytesPerTile, BytesPerTileId.oneK);
        expect(info.mtu, 1028);
        expect(info.symbolSize, 1024);
        expect(info.k, (5000 / 1024).ceil());
        expect(info.totalSize, 5000);
        expect(info.compressedSize, 5000);
        expect(info.compressed, isFalse);
        expect(info.fileSHA256, sha256Hex(file));
        expect(info.sessionId, matches(RegExp(r'^[0-9a-f]{16}$')));
        expect(prepared.metaFrames, hasLength(1));
      },
    );

    test('one DATA frame per source symbol with correct wire fields', () async {
      final prepared = await prepareTransfer(
        file: file,
        filename: 'sample.bin',
        mime: 'application/octet-stream',
        factory: _FakeFactory(),
      );
      final info = prepared.info;

      expect(prepared.dataFrames, hasLength(info.k));
      for (var i = 0; i < prepared.dataFrames.length; i++) {
        final frame = decodeFrame(prepared.dataFrames[i]);
        expect(frame.type, typeData);
        expect(frame.sessionId, info.sessionId);
        expect(frame.esi, i);
        expect(frame.k, info.k);
        expect(frame.totalLen, info.compressedSize);
        expect(frame.flags, 0);
        expect(frame.payload.length, info.symbolSize);
      }
    });

    test('META frame parses to all fields matching TransferInfo', () async {
      final prepared = await prepareTransfer(
        file: file,
        filename: 'sample.bin',
        mime: 'application/octet-stream',
        factory: _FakeFactory(),
      );
      final info = prepared.info;

      final meta = parseMetadataFrame(prepared.metaFrames.single);
      expect(meta.magic, metaMagic);
      expect(meta.protoVer, protoVersion);
      expect(meta.sessionId, info.sessionId);
      expect(meta.filename, 'sample.bin');
      expect(meta.mime, 'application/octet-stream');
      expect(meta.totalSize, info.totalSize);
      expect(meta.compressed, isFalse);
      expect(meta.k, info.k);
      expect(meta.symbolSize, info.symbolSize);
      expect(meta.mtu, info.mtu);
      expect(meta.fileSHA256, info.fileSHA256);
      expect(meta.flags, 0);
    });
  });

  group('metadata compressedSize quirk', () {
    test(
      'pins META compressedSize to 0 while DATA frames carry the true length',
      () async {
        final file = randomBytes(3000);
        final prepared = await prepareTransfer(
          file: file,
          filename: 'u.bin',
          mime: 'application/octet-stream',
          factory: _FakeFactory(),
        );
        final info = prepared.info;

        // Uncompressed: the metadata document and the META wire frame's totalLen
        // both report 0, while every DATA frame carries the true wire length.
        final meta = parseMetadataFrame(prepared.metaFrames.single);
        expect(meta.compressedSize, 0);
        expect(decodeFrame(prepared.metaFrames.single).totalLen, 0);
        expect(
          decodeFrame(prepared.dataFrames.first).totalLen,
          info.compressedSize,
        );
      },
    );
  });

  group('compressible input', () {
    test('compresses and round-trips to the original file', () async {
      final file = Uint8List.fromList(List.filled(10000, 0x41)); // 10000 'A'
      final prepared = await prepareTransfer(
        file: file,
        filename: 'a.txt',
        mime: 'text/plain',
        factory: _FakeFactory(),
      );
      final info = prepared.info;

      expect(info.compressed, isTrue);
      expect(info.compressedSize, lessThan(info.totalSize));

      final decoder = await _FakeFactory().createDecoder(
        info.compressedSize,
        info.mtu,
      );
      Uint8List? recovered;
      for (final frameBytes in prepared.dataFrames) {
        final frame = decodeFrame(frameBytes);
        expect(frame.flags, flagCompressed);
        recovered = decoder.decode(frame.payload);
      }
      expect(recovered, isNotNull);
      expect(
        decompress(
          CompressedResult(data: recovered!, compressed: info.compressed),
        ),
        file,
      );
    });
  });

  group('errors', () {
    test('empty file → emptyFile', () async {
      await expectLater(
        prepareTransfer(
          file: Uint8List(0),
          filename: 'e.bin',
          mime: 'application/octet-stream',
          factory: _FakeFactory(),
        ),
        throwsA(pipelineError(PipelineErrorCode.emptyFile)),
      );
    });

    test('file over the wire totalLen limit → tooLarge', () async {
      final big = randomBytes(maxTotalLen + 1, 7);
      await expectLater(
        prepareTransfer(
          file: big,
          filename: 'big.bin',
          mime: 'application/octet-stream',
          factory: _FakeFactory(),
        ),
        throwsA(pipelineError(PipelineErrorCode.tooLarge)),
      );
    });

    test(
      'encoder symbol size beyond the frame budget → frameTooLarge',
      () async {
        final factory = _FakeFactory(symbolSizeOverride: 2000);
        await expectLater(
          prepareTransfer(
            file: randomBytes(100),
            filename: 'f.bin',
            mime: 'application/octet-stream',
            factory: factory,
          ),
          throwsA(pipelineError(PipelineErrorCode.frameTooLarge)),
        );
      },
    );

    test('encoder k too small for the payload → kSanityFailed', () async {
      final factory = _FakeFactory(forceSourceSymbolCount: 1);
      await expectLater(
        prepareTransfer(
          file: randomBytes(10000),
          filename: 'f.bin',
          mime: 'application/octet-stream',
          factory: factory,
        ),
        throwsA(pipelineError(PipelineErrorCode.kSanityFailed)),
      );
    });

    test('invalid settings → ArgumentError propagates', () async {
      const bad = TransferSettings(
        bytesPerTile: BytesPerTileId.oneK,
        layout: LayoutId.grid4,
        targetFps: 99,
        highRefresh: false,
      );
      await expectLater(
        prepareTransfer(
          file: randomBytes(100),
          filename: 'f.bin',
          mime: 'application/octet-stream',
          settings: bad,
          factory: _FakeFactory(),
        ),
        throwsArgumentError,
      );
    });
  });

  group('repair frames', () {
    test('repairFrames emits count DATA frames with esi >= k', () async {
      final prepared = await prepareTransfer(
        file: randomBytes(3000),
        filename: 'f.bin',
        mime: 'application/octet-stream',
        factory: _FakeFactory(),
      );
      final info = prepared.info;

      final repair = repairFrames(prepared, 3);
      expect(repair, hasLength(3));
      for (final frameBytes in repair) {
        final frame = decodeFrame(frameBytes);
        expect(frame.type, typeData);
        expect(frame.sessionId, info.sessionId);
        expect(frame.esi, greaterThanOrEqualTo(info.k));
        expect(frame.k, info.k);
        expect(frame.totalLen, info.compressedSize);
        expect(frame.payload.length, info.symbolSize);
      }
    });

    test('negative repair count → badRepairCount', () async {
      final prepared = await prepareTransfer(
        file: randomBytes(100),
        filename: 'f.bin',
        mime: 'application/octet-stream',
        factory: _FakeFactory(),
      );
      expect(
        () => repairFrames(prepared, -1),
        throwsA(pipelineError(PipelineErrorCode.badRepairCount)),
      );
    });
  });

  group('determinism', () {
    test(
      'same input twice → identical symbol payloads (deterministic fake)',
      () async {
        final file = randomBytes(4000);
        final first = await prepareTransfer(
          file: file,
          filename: 'a.bin',
          mime: 'application/octet-stream',
          factory: _FakeFactory(),
        );
        final second = await prepareTransfer(
          file: file,
          filename: 'a.bin',
          mime: 'application/octet-stream',
          factory: _FakeFactory(),
        );

        // Each send session gets a fresh sessionId, but the fake factory is
        // deterministic, so the DATA symbol payloads are byte-identical.
        expect(first.info.sessionId, isNot(second.info.sessionId));
        expect(second.dataFrames, hasLength(first.dataFrames.length));
        for (var i = 0; i < first.dataFrames.length; i++) {
          final f1 = decodeFrame(first.dataFrames[i]);
          final f2 = decodeFrame(second.dataFrames[i]);
          expect(f1.esi, i);
          expect(f2.esi, i);
          expect(f2.payload, f1.payload);
          expect(f2.k, f1.k);
          expect(f2.totalLen, f1.totalLen);
          expect(f2.flags, f1.flags);
        }
      },
    );
  });
}
